### PR TITLE
Add capitalize_first method

### PR DESCRIPTION
**Pull request by glaucocustodio**
Ruby's  `capitalize` method turns all characteres in lowercase, except the first one, which turns uppercase. The `capitalize_first` method turns the first character uppercase and keeps the others as it is.

`"what a Lovely Day".capitalize # => "What a lovely day"`

`"what a Lovely Day".capitalize_first # => "What a Lovely Day"`

If another method is not good, we can add an option to `capitalize` maybe.

### CODE REVIEW

File: activesupport/lib/active_support/core_ext/string/inflections.rb [hunk 1]
```diff
@@ -222,6 +222,13 @@ def humanize(options = {})
     ActiveSupport::Inflector.humanize(self, options)
   end
 
+  # Converts just the first character to uppercase.
+  #
+  #  'what a Lovely Day'.capitalize # => "What a Lovely Day"
```
Comment (tadeusrox) [inline activesupport/lib/active_support/core_ext/string/inflections.rb]: `'what a Lovely Day'.capitalize` => `'what a Lovely Day'.capitalize_first` ?

Comment (glaucocustodio) [inline activesupport/lib/active_support/core_ext/string/inflections.rb]: Thank you, I've fixed.

File: activesupport/lib/active_support/inflector/methods.rb [hunk 2]
```diff
@@ -140,6 +140,10 @@ def humanize(lower_case_and_underscored_word, options = {})
       result
     end
 
+    def capitalize_first string
```
Comment (egilburg) [inline activesupport/lib/active_support/inflector/methods.rb]: convention is to use parentheses in method names, `def capitalize_first(string)`

**Comment by rafaelfranca** [comment]
@fxn what do you think?

**Comment by egilburg** [comment]
:+1: on the feature, we had this requirement in our codebase too, implemented as:

```ruby
  self[0].upcase + self[1..-1]
```

Here's a benchmark:

```ruby
class String
  def upcase_range
    self[0].upcase + self[1..-1]
  end

  def upcase_sub
    sub(/\S/, &:upcase)
  end
end

require 'benchmark'

n = 1_000_000
str = 'what a Lovely Day'

Benchmark.bmbm do |x|
  x.report('#upcase_range') do
    n.times { str.upcase_range }
  end

  x.report('#upcase_sub') do
    n.times { str.upcase_sub }
  end
end
```

Result:

```
Rehearsal -------------------------------------------------
#upcase_range   0.390000   0.000000   0.390000 (  0.393900)
#upcase_sub    1.380000   0.000000   1.380000 (  1.380069)
---------------------------------------- total: 1.770000sec

                    user     system      total        real
#upcase_range   0.390000   0.000000   0.390000 (  0.388639)
#upcase_sub    1.420000   0.010000   1.430000 (  1.419239)
```

Seems the index+range implementation is about 3.5 times faster than with sub.

**Comment by akshaymohite** [comment]
Here is `benchmark-ips` report showing memory footprint as well.
```
Calculating -------------------------------------
 #upcase_first_index   101.810k i/100ms
   #upcase_first_sub    54.490k i/100ms
-------------------------------------------------
 #upcase_first_index      2.272M (± 8.0%) i/s -     11.301M
   #upcase_first_sub    779.316k (± 7.1%) i/s -      3.923M
```

**Comment by glaucocustodio** [comment]
Given that string:  `" what a Lovely Day"`

How would we deal with it? Shall we turn `w` uppercase? `sub(/\S/, &:upcase)` does it, ` self[0].upcase + self[1..-1]` does not.

File: activesupport/lib/active_support/inflector/methods.rb (hunk 2 above)
Comment (glaucocustodio) [inline activesupport/lib/active_support/inflector/methods.rb]: Thanks.

**Comment by egilburg** [comment]
@akshaymohite could you do a footprint on variation of `self[0].upcase.concat(self[1..-1])` ?

**Comment by egilburg** [comment]
@glaucocustodio is that really the role of capitalize_first to deal with blank characters? If string has blank spaces IMO it should first be squished.

**Comment by glaucocustodio** [comment]
I think the same as you @egilburg, that was only to point out a performance/utility tradeoff.

**Comment by egilburg** [comment]
Some more implementations :)

```ruby
class String
  def upcase_first_index_interpolate
    "#{self[0].upcase}"#{self[1..-1])}"
  end

  def upcase_first_index_concat
    self[0].upcase.concat(self[1..-1])
  end

  def upcase_first_index_plus
    self[0].upcase + self[1..-1]
  end

  def upcase_first_dup_replace
    dup = self.dup
    dup[0] = dup[0].upcase
    dup
  end

  def upcase_first_sub
    sub(/\S/, &:upcase)
  end

  def upcase_first_dup_sub
    dup.sub!(/\S/, &:upcase)
  end
end

require 'benchmark'
require 'benchmark/ips'

str = 'what a Lovely Day'.freeze

Benchmark.ips do |x|
  x.config(:time => 5, :warmup => 1)

  x.report('#upcase_first_index_interpolate') do
    str.upcase_first_index_interpolate
  end

  x.report('#upcase_first_index_concat') do
    str.upcase_first_index_concat
  end

  x.report('#upcase_first_index_plus') do
    str.upcase_first_index_plus
  end

  x.report('#upcase_first_dup_replace') do
    str.upcase_first_dup_replace
  end

  x.report('#upcase_first_sub') do
    str.upcase_first_sub
  end

  x.report('#upcase_first_dup_sub') do
    str.upcase_first_dup_sub
  end
end
```

Results:

```
Calculating -------------------------------------
#upcase_first_index_interpolate
                          2.846M (± 1.2%) i/s -     14.250M
#upcase_first_index_concat
                          2.121M (± 2.6%) i/s -     10.641M
#upcase_first_index_plus
                          2.117M (± 2.1%) i/s -     10.615M
#upcase_first_dup_replace
                          1.547M (± 1.9%) i/s -      7.776M
   #upcase_first_sub    690.771k (± 1.8%) i/s -      3.491M
#upcase_first_dup_sub
                        542.632k (± 1.4%) i/s -      2.735M
```

Seems a pretty consistent time-space tradeoff. Given it's linearity, I'd argue most would prioritize time over memory usage, as memory can be increased (and is garbage collected after use) while time is paid by the user or a more complex parallelization algorithm.

**Comment by fxn** [comment]
Which use cases does this method have?

**Comment by egilburg** [comment]
@fxn Similar as `capitalize` but without destroying existing capital case of proper nouns. Used for list items where it would be beginning of sentence and thus should start capitalized.

```ruby
"user Joe".capitalize # => "User joe"
"user Joe".capitalize_first # => "User Joe"
```

File: activesupport/test/inflector_test.rb [hunk 3]
```diff
@@ -157,6 +157,10 @@ def test_acronyms
     end
   end
 
+  def test_capitalize_first
+    assert_equal "What a Lovely Day", ActiveSupport::Inflector.capitalize_first("what a Lovely Day")
```
Comment (egilburg) [inline activesupport/test/inflector_test.rb]: please also add test for the delegation, `assert_equal "What a Lovely Day", "what a Lovely Day".capitalize_first`, as that will likely be more common usage than manually using the inflector.

**Comment by matthewd** [comment]
> If another method is not good, we can add an option to capitalize maybe.

I suspect this would be the better choice.

It seems arbitrary that `capitalize_first` means "make the first letter a capital and leave the rest", and `capitalize` means "make the first letter a capital and lowercase the rest": if anything, I think I'd expect them to be the other way around.

A kwarg option allows us to describe the actual difference in behaviour.

**Comment by fxn** [comment]
I would like to understand use cases, but the name is something that does not click to me in either case.

The definition of `capitalize` has two parts: what happens to the first character (upcased), and what happens to the rest (lowercased). It could be argued that by saying "first" the name implies "rest" is an empty string and thus logically anything applies to an empty string. But... doesn't click.

The name that would really make sense to me would be `upcase_first`, just upcase the first character (vs upcase all).

**Comment by egilburg** [comment]
:+1: for `upcase_first`, that's how our implementation is named as well.

**Comment by glaucocustodio** [comment]
`upcase_first` seems better to me, simpler and shorter. I have renamed the method and changed its implementation to the faster one.

**Comment by rafaelfranca** [comment]
r? @fxn 

@glaucocustodio could you rebase it?

**Pull request closed without merging**